### DISCUSSION TITLE
ESP32 TinyUSB Keyboard does not work at boot

### DISCUSSION THREAD

**Discussion by Jackolix** [hathach/tinyusb#2795] [Q&A]
Hey,

im using TinyUSB on my Espressif ESP32s3 to send keyboard commands to a connected PC.

For testing im using a Lenovo Thinkpad, where you have to press ENTER at Boot to get to the BOOT-menu. Sadly the pc doesnt recognize the ENTER spamming from the ESP. In the Boot-menu itself it does recognize ESP keyboard inputs.

Do i have to configure something extra to get it work at boot or something different?
C code which is running on the esp is here: https://pastes.dev/CNzI2lWa6g
I did a lot of testing so it is not the cleanest code.

Any help appreciated!

**Comment by Jackolix** [comment]
I rewrote my program just to send CAPS LOCK and see if the computer responds. Which he does on an old DELL Tower pc but not on the Lenovo thinkpad.

Maybe Lenovo does only load spesific devices which are from a big manufacturer at boot? I have no clue.
Maybe my tinyusb descriptors are wrong?

**Comment by HiFiPhile** [comment]
I'm not familiar with ESP wrapper layer, I suppose the BIOS has limited USB support where only certain kinds of devices are recognized.
You can dump the descriptors of a real mouse and compare with your device, with tools like [usbtreeview](https://www.uwe-sieber.de/usbtreeview_e.html)

**Reply by Jackolix** [reply]
I've already attempted that approach and modified the descriptors to be nearly identical to those of a generic USB keyboard. Unfortunately, I couldn't alter the bMaxPacketSize0 and wDescriptorLength values without causing issues with the esp's functionality.
![Screenshot 2024-09-16 104236](https://github.com/user-attachments/assets/f7889bbc-2f50-4560-aeef-72247d389b3f)

**Comment by HiFiPhile** [comment]
You can use TinyUSB without ESP wrapper layer, which is not covered by this repo.

https://github.com/hathach/tinyusb/tree/master/examples/device/hid_boot_interface
